src/wayland/dmabuf/mod.rs

```diff
@@ -112,18 +119,311 @@ use crate::{
+    utils::{ids::id_gen, sealed_file::SealedFile, UnmanagedResource},
 };
 
-use super::buffer::BufferHandler;
+use super::{buffer::BufferHandler, compositor};
+
+#[derive(Debug, Clone, PartialEq)]
+struct DmabufFeedbackTranche {
+    target_device: libc::dev_t,
+    flags: zwp_linux_dmabuf_feedback_v1::TrancheFlags,
+    indices: IndexSet<usize>,
+}
+
+#[derive(Debug)]
+struct DmabufFeedbackFormatTable {
+    formats: IndexSet<Format>,
+    file: SealedFile,
+}
+
+#[derive(Debug)]
+struct DmabufFeedbackInner {
+    main_device: libc::dev_t,
+    format_table: DmabufFeedbackFormatTable,
+    tranches: Vec<DmabufFeedbackTranche>,
+}
+
+impl PartialEq for DmabufFeedbackInner {
+    fn eq(&self, other: &Self) -> bool {
+        self.main_device == other.main_device
+            && self.format_table.formats == other.format_table.formats
+            && self.tranches == other.tranches
+    }
+}
+
+/// Builder for [`DmabufFeedback`]
+#[derive(Debug, Clone)]
+pub struct DmabufFeedbackBuilder {
+    main_device: libc::dev_t,
+    main_tranche: DmabufFeedbackTranche,
+    formats: IndexSet<Format>,
+    preferred_tranches: Vec<DmabufFeedbackTranche>,
+}
+
+#[repr(C, packed)]
+struct DmabufFeedbackFormat {
+    format: u32,
+    _reserved: u32,
+    modifier: u64,
+}
+
+impl From<Format> for DmabufFeedbackFormat {
+    fn from(format: Format) -> Self {
+        DmabufFeedbackFormat {
+            format: format.code as u32,
+            _reserved: 0,
+            modifier: format.modifier.into(),
+        }
+    }
+}
+
+impl DmabufFeedbackBuilder {
+    /// Create a new feedback builder with the specified device and formats as the main device
+    ///
+    /// Preferred tranches can be added with [`DmabufFeedbackBuilder::add_preferred_tranche`]
+    /// and the main tranche will be put after all preferred tranches
+    pub fn new(main_device: libc::dev_t, formats: impl IntoIterator<Item = Format>) -> Self {
+        let feedback_formats: IndexSet<Format> = formats.into_iter().collect();
+        let format_indices: IndexSet<usize> = (0..feedback_formats.len()).collect();
+        let main_tranche = DmabufFeedbackTranche {
+            flags: zwp_linux_dmabuf_feedback_v1::TrancheFlags::empty(),
+            indices: format_indices,
+            target_device: main_device,
+        };
+
+        Self {
+            main_device,
+            formats: feedback_formats,
+            main_tranche,
+            preferred_tranches: Vec::new(),
+        }
+    }
+
+    /// Adds a preferred tranche to the builder
+    ///
+    /// The tranches will be reported in the order they have been added with
+    /// this function.
+    ///
+    /// Note: Formats already present in a previously added preferred tranche with the
+    /// same target device and flags will be skipped.
+    /// If all formats are already included in a previously added tranche
+    /// with the same target device and flags this tranche will be skipped.
+    pub fn add_preferred_tranche(
```

Comment:
"Preference tranche" is the term used in the protocol spec, rather than "preferred tranche".

Reply:
renamed to preference tranche